test/unit/pipeline/FetchUnitTest.cc

```diff
@@ -442,5 +462,288 @@ TEST_F(PipelineFetchUnitTest, idleLoopBufferDueToNotTakenBoundary) {
+  ON_CALL(isa, getMaxInstructionSize()).WillByDefault(Return(insnMaxSizeBytes));
+  ON_CALL(isa, getMinInstructionSize()).WillByDefault(Return(insnMinSizeBytes));
+  ON_CALL(memory, getCompletedReads()).WillByDefault(Return(completedReads));
+
+  // Buffer will contain valid min size instruction so predecode returns
+  // min bytes read
+  MacroOp mOp = {uopPtr};
+  ON_CALL(isa, predecode(_, insnMinSizeBytes, 0x10 - insnMinSizeBytes, _))
+      .WillByDefault(DoAll(SetArgReferee<3>(mOp), Return(insnMinSizeBytes)));
+
+  // Fetch the data, only min bytes will be copied to fetch buffer. Should allow
+  // continuation to predecode
+  EXPECT_CALL(isa, getMaxInstructionSize()).Times(1);
+  EXPECT_CALL(memory, getCompletedReads()).Times(1);
+  EXPECT_CALL(isa, getMinInstructionSize()).Times(1);
+  EXPECT_CALL(isa, predecode(_, _, _, _)).Times(1);
+
+  uint64_t setPC = blockSize - insnMinSizeBytes;
+  // Fetch a single minimum sized instruction, buffered bytes = 0
+  fetchUnit.updatePC(setPC);
+  // Tick. Fetch data and predecode
+  fetchUnit.tick();
+
+  // Buffer should now be empty as all bytes predecoded
+  EXPECT_EQ(fetchUnit.bufferedBytes_, 0);
+  EXPECT_EQ(fetchUnit.output_.getTailSlots()[0], mOp);
+
+  // Expect a block starting at address 16 to be requested when we fetch again
+  EXPECT_CALL(isa, getMaxInstructionSize()).Times(1);
+  EXPECT_CALL(memory,
+              requestRead(Field(&memory::MemoryAccessTarget::address, 16), _))
+      .Times(1);
+  fetchUnit.requestFromPC();
+
+  // Tick again, expecting that decoding will now resume
+  MacroOp mOp2 = {uopPtr2};
+  memory::MemoryReadResult nextBlockValue = {{16, blockSize}, 0, 1};
+  span<memory::MemoryReadResult> nextBlock = {&nextBlockValue, 1};
+  ON_CALL(memory, getCompletedReads()).WillByDefault(Return(nextBlock));
+  ON_CALL(isa, predecode(_, _, _, _))
+      .WillByDefault(DoAll(SetArgReferee<3>(mOp2), Return(insnMaxSizeBytes)));
+
+  EXPECT_CALL(isa, getMaxInstructionSize()).Times(1);
+  // Completed reads called again as more data is requested
+  EXPECT_CALL(memory, getCompletedReads()).Times(1);
+  EXPECT_CALL(isa, getMinInstructionSize()).Times(1);
+  // Output of width 1 so only 1 call to predecode
+  EXPECT_CALL(isa, predecode(_, _, _, _)).Times(1);
+
+  fetchUnit.tick();
+
+  // Initially 0 bytes, 16 bytes added, max bytes predecoded leaving (16 - max)
+  // bytes left
+  EXPECT_EQ(fetchUnit.bufferedBytes_, 16 - insnMaxSizeBytes);
+  EXPECT_EQ(fetchUnit.output_.getTailSlots()[0], mOp2);
+}
+
+// Test that invalid min number of bytes held at the end of the buffer is not
+// successfully predecoded and that more data is fetched subsequently allowing
+// progression as a full instruction is now present in the buffer
+TEST_P(PipelineFetchUnitTest, invalidMinBytesAtEndOfBuffer) {
+  // This is only relevant if min and max size are different. Otherwise, there
+  // won't be any progression as the fetch unit will be caught in an infinite
+  // loop
```

Comment:
If we get stuck in an infinite loop for the following scenario when insnMinSize and insnMaxSize are equal then is this not a bug that needs fixing?

Reply:
Is this even a possible state to achieve when they are equal? It would require insnMinSize bytes to be in the fetch buffer, which is certainly an instruction or eof in our case, so you wouldn't be stuck in a case with this number of bytes but not an instruction in the buffer. Please correct me if I'm wrong

Reply:
Invalid as in predecode returns 0. The only (intended) case of this is where RISC-V predecode thinks it needs 4 bytes but has only been given 2. It is returning 0 to say it needs more data. With min and max the same predecode will always consume 4 bytes